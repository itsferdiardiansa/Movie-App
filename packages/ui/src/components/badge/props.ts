export type BadgeVariant =
  | 'primary'
  | 'light'
  | 'danger'
  | 'warning'
  | 'dark'
  | 'success'
  | 'orange'

export type BadgeSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl'

export type BadgeShape = 'rounded' | 'pill' | 'square'

export interface BadgeTypes {
  label: string
  variant: BadgeVariant
  size: BadgeSize
  shape: BadgeShape
  inverse: boolean
  icon: () => string[]
  dot: boolean
}

export const defaultProps: BadgeTypes = {
  label: '',
  variant: 'light',
  size: 'xs',
  shape: 'rounded',
  inverse: false,
  icon: () => [],
  dot: false,
}
