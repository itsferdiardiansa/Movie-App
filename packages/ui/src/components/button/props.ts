export type ButtonVariant =
  | 'primary'
  | 'light'
  | 'danger'
  | 'warning'
  | 'dark'
  | 'success'
  | 'orange'

export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl'

export type ButtonShape = 'rounded' | 'pill' | 'square'

export interface ButtonTypes {
  label: string
  variant: ButtonVariant
  size: ButtonSize
  shape: ButtonShape
  disabled: boolean
  isLoading: boolean
  block: boolean
  inverse: boolean
}

export const defaultProps: ButtonTypes = {
  label: 'Click me!',
  variant: 'primary',
  size: 'md',
  block: false,
  inverse: false,
  disabled: false,
  shape: 'rounded',
  isLoading: false,
}
