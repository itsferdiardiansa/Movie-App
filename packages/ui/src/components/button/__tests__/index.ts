import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import { Button } from '..'
import { defaultProps } from '../props'

describe('Components/Button', () => {
  it('renders correctly (snapshot)', () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Test Button',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: false,
        inverse: false,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the button with default props', () => {
    const wrapper = mount(Button, {
      props: {
        label: defaultProps.label,
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: false,
        inverse: false,
      },
    })
    const button = wrapper.find('button')

    expect(button.exists()).toBe(true)
    expect(button.text()).toBe(defaultProps.label)
    expect(button.classes()).toContain('tx-btn')
    expect(button.classes()).toContain('tx-btn--primary')
    expect(button.classes()).toContain('tx-btn--md')
  })

  it('renders the button with a custom label', () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Submit',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: false,
        inverse: false,
      },
    })
    expect(wrapper.find('button').text()).toBe('Submit')
  })

  it('emits click event with payload when clicked', async () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Click Me',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: false,
        inverse: false,
      },
    })
    const button = wrapper.find('button')
    await button.trigger('click')

    expect(wrapper.emitted()).toHaveProperty('click')
    const clickEvent = wrapper.emitted('click')
    expect(clickEvent).toHaveLength(1)
    expect(clickEvent?.[0]).toBeTruthy()
  })

  it('does not emit click event when disabled', async () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Submit',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: true,
        isLoading: false,
        block: false,
        inverse: false,
      },
    })
    const button = wrapper.find('button')
    await button.trigger('click')
    expect(wrapper.emitted('click')).toBeFalsy()
  })

  it('applies correct classes for variant and size', () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Submit',
        variant: 'danger',
        size: 'lg',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: false,
        inverse: false,
      },
    })
    const button = wrapper.find('button')
    expect(button.classes()).toContain('tx-btn--danger')
    expect(button.classes()).toContain('tx-btn--lg')
  })

  it('shows spinner and hides label when isLoading is true', () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Submit',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: true,
        block: false,
        inverse: false,
      },
    })

    expect(wrapper.find('.tx-btn--spinner').exists()).toBe(true)
    expect(wrapper.find('.tx-btn--label').exists()).toBe(false)
  })

  it('applies block class when block prop is true', () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Submit',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: true,
        inverse: false,
      },
    })
    expect(wrapper.find('button').classes()).toContain('block')
  })

  it('applies correct class based on shape prop', () => {
    const shapes = ['pill', 'square', 'rounded'] as const
    shapes.forEach(shape => {
      const wrapper = mount(Button, {
        props: {
          label: 'Submit',
          variant: 'primary',
          size: 'md',
          shape,
          disabled: false,
          isLoading: false,
          block: false,
          inverse: false,
        },
      })
      expect(wrapper.find('button').classes()).toContain(shape)
    })
  })

  it('renders custom content in default slot', () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Submit',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: false,
        inverse: false,
      },
      slots: {
        default: '<span class="custom-content">Custom Button</span>',
      },
    })
    expect(wrapper.find('.custom-content').exists()).toBe(true)
    expect(wrapper.find('.custom-content').text()).toBe('Custom Button')
  })

  it('applies the inverse variant class', () => {
    const wrapper = mount(Button, {
      props: {
        label: 'Submit',
        variant: 'primary',
        size: 'md',
        shape: 'rounded',
        disabled: false,
        isLoading: false,
        block: false,
        inverse: true,
      },
    })
    expect(wrapper.find('button').classes()).toContain(
      'tx-btn--primary-inverse'
    )
  })
})
