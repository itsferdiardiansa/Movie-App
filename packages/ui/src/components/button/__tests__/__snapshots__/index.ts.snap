// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components/Button > renders correctly (snapshot) 1`] = `
"<button class="tx-btn rounded tx-btn--primary tx-btn--md">
  <div class="tx-btn--wrapper">
    <div class="tx-btn--label">Test Button</div>
  </div>
</button>"
`;
